### PR TITLE
[Logs][Fix] Fix logs tagging inconsistency for restarted containers

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

- Consider the container ID in the file tailers map by using the key format `<filePath/containerID>` instead of `<filePath>`
- Ensure thread-safety

### Motivation

On Kubernetes we use wildcard to match container log files (e.g `/var/log/pods/<ns>_<pod_name>_<pod_uid>/<container_name>/*.log`, `*` is an integer).
When a container restarts a new container ID is generated and a new log file appears with the path `/var/log/pods/<ns>_<pod_name>_<pod_uid>/<container_name>/n+1.log`.

The logs agent relies on AD to get this kind of service updates (a container restart in AD is a container dead + new container created), but the logs agent also tries to scan the current configs it has every 10 seconds and if it starts a scan after the creation of the file `n+1.log` and before getting the update from AD (which happens quite often) the new log file of the new container will match the log config of the dead container because of the wildcard and the logs agent will start collecting logs of the new container using the old container's ID.

Few seconds later AD sends the info but the logs agent will ignore collecting logs from the new container because in its cache it has already a (wrong) tailer for that file. And what is even worse is that at some point the Tagger expires the dead container's tags and the logs agent will keep sending logs without Tagger tags because it uses a dead container's ID to fetch tags.

Even though this PR makes it at least impossible to block the logs collection from a restarted container with the new/correct container ID, because of the racy behaviour + the wildcard config explained above the logs agent will keep collecting new container's logs with the dead container ID for few seconds until the AD update comes, this result in having duplicated logs for few seconds, **and it could be a big amount of duplicated logs depending on the logging frequency and the delay between AD update and the new scan.**

### Important Notes

This a short-term fix, as we prefer having duplicated logs for few seconds instead of not having logs at all or having logs with wrong container ID and missing Tagger tags.

### Describe your test plan

Force a container restart `k exec <pod_name> -- /bin/sh -c "kill 1"` and look at the log explorer.

### CODE REVIEW

File: pkg/logs/input/file/file_provider.go
Comment (clamoriniere) [inline pkg/logs/input/file/file_provider.go:49]: I think it is better to check for nil pointer because `Source` and `Config` can be `nil`

File: pkg/logs/input/file/scanner.go
Comment (clamoriniere) [inline pkg/logs/input/file/scanner.go:38]: if `sync.Mutex` is a composition I think it is better to be on the top of the struct field.

it feel like the mutex is only use to protect the `tailers` map, have you considered to name the mutex according the the map name like `tailersMutex` ?

File: pkg/logs/input/file/scanner.go
Comment (clamoriniere) [inline pkg/logs/input/file/scanner.go:93]: the mutex should be `Lock()` and `Unlock()` in this function.

File: pkg/logs/input/file/scanner.go
Comment (clamoriniere) [inline pkg/logs/input/file/scanner.go:163]: forget about my previous comment, the Mutex is also used for the `activeSources` slice.

**Review by clamoriniere — COMMENTED** [review]
Thanks @ahmed-mez for the fix, some small comments

File: pkg/logs/input/file/file_provider.go
Comment (ahmed-mez) [reply pkg/logs/input/file/file_provider.go:49]: we're making assumptions in the code that it's not nil before even calling this method, but I'll check for nil pointers here as you suggested 👍

**Review by xornivore — APPROVED** [review]
🚀

**Review by olivielpeau — COMMENTED** [review]
Thanks for the investigation and the fix 🙇 

I think this makes sense overall, and indeed it's tricky to fix this cleanly. A couple of thoughts on the current code:

1. I don't think there's need for a mutex, since all the operations on the `Scanner` struct fields are performed from the same goroutine: the one that runs `Scanner.run`: https://github.com/DataDog/datadog-agent/blob/4a8537c563bf9cd6bd99a6e9610ece04a3650ec3/pkg/logs/input/file/scanner.go#L69-L87 Am I missing something though?
2. I'm wondering if this new "tailer key" should also be used as the messages' `Origin.Identifier` set here: https://github.com/DataDog/datadog-agent/blob/4a8537c563bf9cd6bd99a6e9610ece04a3650ec3/pkg/logs/input/file/tailer.go#L218 . Thinking of the impact on registered offsets in particular. I'll think about this again.

**Comment by ahmed-mez** [comment]
Thanks for looking !
1. It's because we have pointers to tailers in the scanner's store, those tailers spawn other goroutines and by locking the scanner's store when run a scan we ensure the tailers and files we reference aren't being concurrently altered by their own goroutines.
I can confirm that with the mutex I can see a more stable behaviour (surprisingly without the mutex we loose few first logs from the new restarted container's sometimes, maybe it's another problem we should look into?).
2. TBH I don't know the role of `origin.Identifier` and the impact of changing it, I can make the change if you think we should.

File: pkg/logs/input/file/tailer_key.go
Comment (remeh) [inline pkg/logs/input/file/tailer_key.go:11]: About this interface, why not only having a `Key() string` method? Do you have other plans for `getPath()` and `getSourceIdentifier()`?

That would not avoid the extra `buildTailerKey()` function but simplify the role of the interface.

**Review by remeh — COMMENTED** [review]
Your change looks good to me and smart enough to not be as "short-term" as you mentioned in the PR description because I think that the actual fix would need such a key. Fixing this "long-term" will be touchy but, if we can accept having a 10 seconds delay (assuming the scan is done <= every 10 sec), it should be doable without completely changing the scanner implementation, using temporary sources.
About the first point mentioned by @olivielpeau, I would say that it is because of the `activeSources` which may not be consistent without the mutex, but not 100% sure.

File: pkg/logs/input/file/tailer_key.go
Comment (ahmed-mez) [reply pkg/logs/input/file/tailer_key.go:11]: yes I considered that option and preferred to keep the key generating logic isolated and in one place not duplicated in the file_provider and the tailer.
I thought it's better this way because we want to make sure file_provider and tailer always return the same key, in other words, it safer to let the scanner build its own map keys and IMHO more robust to future changes in file_provider and tailer.
wdyt?

File: pkg/logs/input/file/tailer_key.go
Comment (ahmed-mez) [reply pkg/logs/input/file/tailer_key.go:17]: I'm not sure I understood the point, are you saying we can drop the `<path>/` prefix?

File: pkg/logs/input/file/tailer_key.go
Comment (remeh) [reply pkg/logs/input/file/tailer_key.go:11]: Sounds totally sensible! 👍

File: pkg/logs/input/file/tailer_key.go
Comment (remeh) [reply pkg/logs/input/file/tailer_key.go:17]: I'm wondering what would happen in the case of Tailer (or File) returning an empty string, because that branch exist: https://github.com/DataDog/datadog-agent/pull/5745/files#diff-de8b9f30f737ed4f550714585cacee54R52
It will probably never happen if the NewTailer and NewFile constructor is properly used everywhere.

Comment (ahmed-mez) [reply pkg/logs/input/file/tailer_key.go:17]: yes in that case we will end up with a `<path>/` prefix, similar to what we had before this PR

Comment (ahmed-mez) [reply pkg/logs/input/file/tailer_key.go:17]: it will reproduce the same behaviour that this PR tries to fix

**Review by remeh — APPROVED** [review]

**Review by olivielpeau — COMMENTED** [review]
Thanks for your responses:

1. about the mutex: this is something we should look into more. I don't think we should add a mutex unless we understand exactly the concurrent read/writes we're trying to address. From the code I've read I still don't understand why such a mutex is actually needed by the code. The `tailers` slice only has read/writes from in a single goroutine, and the write operations on each tailer are done atomically I believe.
2. I think `origin.Identifier` is used by the auditor to track offsets. I believe we can leave it to the file path.

One more general question to understand the design of this solution: when the logs agent `scan` logic detects a file rotation on a log file, if it happens before the AD detects that a container has restarted, is there any way for the Logs Agent to know that this rotation was caused by a restarted container?

@remeh let's create a card in backlog to think about this design again

**Comment by ahmed-mez** [comment]
1. More details about the racy behaviour with and without the mutex:
- with the mutex: never miss logs from the new file, tagged with the right container ID, but still receive duplicated logs from the new file tagged with the dead container ID for some time
- without the mutex: miss first logs from the new file when you filter by the new container ID, those missing logs (and more) are collected with the dead container ID also (this is very risky for short lived containers, you can miss all the container's logs if it dies before the new file tailer's start)
I believe it's a race between which tailer starts collecting logs from the new file first, with the mutex we guarantee we never miss logs with the correct container ID, without mutex we can miss logs with the right container ID.
I don't know which concurrent access is causing this though, could it be just the delay caused by the mutex that result in the more stable behaviour? (i.e the locking plays the role of a sleep) not sure though, it's just a hypothesis

**Comment by ahmed-mez** [comment]
More on this 
> is there any way for the Logs Agent to know that this rotation was caused by a restarted container?

log files are actually symlinks to docker files and those docker files have the container ID in the file names and every container has a directory of log files so we can identify container corresponding to the log file, here is example after a container restart
```
root@datadog-rb52v:/# ls -la /var/log/pods/default_logapp_6017607e-3794-462d-bbcd-4415e3754577/logapp/
total 16
drwxr-xr-x 2 root root 4096 Jun 18 12:00 .
drwxr-xr-x 3 root root 4096 Jun 18 12:00 ..
lrwxrwxrwx 1 root root  165 Jun 18 12:00 0.log -> /var/lib/docker/containers/4872d9bb4b25094330195fd6c43621ce929d89a15e21cafd0876226efb188cde/4872d9bb4b25094330195fd6c43621ce929d89a15e21cafd0876226efb188cde-json.log
lrwxrwxrwx 1 root root  165 Jun 18 12:00 1.log -> /var/lib/docker/containers/a4fff86f00e1f319848bd5b1236532a689225eef47d0edaa04a8c466a14ac51a/a4fff86f00e1f319848bd5b1236532a689225eef47d0edaa04a8c466a14ac51a-json.log
```
- dead container ID `4872d9bb4b25094330195fd6c43621ce929d89a15e21cafd0876226efb188cde`
- new container ID  `a4fff86f00e1f319848bd5b1236532a689225eef47d0edaa04a8c466a14ac51a`

I guess this could help for a better design later

**Review by olivielpeau — COMMENTED** [review]
I haven't actually tested the code, but after looking more into how offsets are tracked in the registry, it's less and less clear to me how your PR can actually work reliably. Let me explain:

Offsets of files are tracked by file name (`tailer.Identifier()`). So, with your code, when the scanner starts a new tailer for a new container on a file that's already being tailed by the tailer of the "old" container, the offset in the registry for that file has high chances of both existing in the registry and being the offset of the "old" container's tailer. And then I think the new tailer will use this existing offset, instead of starting from the beginning of the file as we'd want it to.

To solve this, we may want to change the registry so that it tracks file offsets by what `tailer.getSourceIdentifier` returns instead of `tailer.Identifier()`. That said we'll have to make sure we keep backward-compatibility so that when the Agent is upgraded, it still understands the offsets stored in the registry file (but Agent downgrades would still not work well in that regard, not sure if that justifies a new major registry version...).

**Comment by ahmed-mez** [comment]
Makes sense, thank you for the context that's helpful! I think we have two options now
- Merge the PR as temporary fix to mitigate the problem (fwiw the behaviour I described [in this comment](https://github.com/DataDog/datadog-agent/pull/5745#issuecomment-645946619) is consistent)
- Close the PR and let agent core investigate and build a clean solution that covers everything

How can we go forward?  (freeze is next week)

**Comment by olivielpeau** [comment]
Let's keep this open, could you share the specifics of your test setup with @remeh?

**Comment by ahmed-mez** [comment]
yes sure, will sync with him offline 👍

**Comment by remeh** [comment]
@olivielpeau @ahmed-mez 

I think that we basically have this issue because we're using the symlinked logfile as a key in the registry / tailer map.
A solution could be to start resolving the symlinks when the log files are not regular files, and use the resolved file name as the key. This way we would not have collisions (many symlinks can point to the same file) anymore:

  - the resolved filename is containing the container ID (with Docker at least)
  - the key isn't "virtual" anymore, it represents an actual file on disk and creating another entry would just be wrong / a duplication.
 
That's a detail but semantically, the `file:filepath` would keep its sense.

It is not a breaking change for the registry, however, we would send duplicated logs for some time (until every known file has rotated) because of that. To avoid the duplication of logs, we would need to bump the registry version, but we would be able to have a "migration" process trying to resolve the path (if symlinks) before considering them as tailer of the registry v3. If we consider a registry v3, we should probably keep the symlink filepath in the registry entry (for troubleshooting purpose).

(About the mutex, I'm still looking into why it is improving the situation in this case)

WDYT?

**Comment by ahmed-mez** [comment]
Closing this PR, a more robust fix is coming up soon.

**Comment by remeh** [comment]
This best-effort implemented by @ahmed-mez will help attaching the proper tags to logs of a freshly spawned container (within a pod). Still, in a few occasions, we may be missing the very first lines of the logs because the registry will already contain an offset for this file. However, we can't change the keys stored in the registry (today it's only the file path) because if we starts using file path+container_id, we would have every logs file in the directory completely re-tailed for the new container.
For now, let's move forward and merge this PR for 7.23.0, with the plan of improving the situation for next releases, especially if Kubernetes starts adding a feature to be notified on new log file / log rotation.

cc @ahmed-mez @clamoriniere @olivielpeau

**Review by olivielpeau — COMMENTED** [review]
> However, we can't change the keys stored in the registry (today it's only the file path) because if we starts using file path+container_id, we would have every logs file in the directory completely re-tailed for the new container.

Thanks for checking this, and the explanation. I think this is a limitation that could be improved in the future, without waiting for improvements on k8s' side, WDYT? (if you agree, can you track this as a future improvement, and add any thoughts you may already have on how to improve this?).

Overall, after reading the code closely, I still have a few questions/concerns since the logs agent codebase seemed to make a strong assumption up to now that one log file can only have one tailer at a time:

* We focused on the impact of this change on tailed containers, but not really on regular tailed log files. With this change, I guess we could now have more than one tailer for a given log file if this log file is configured to be tailed from more than one integration config file, or am I missing something? This sounds like an undesirable change that could double (or more) the amount of logs sent in these cases.
* `Tailer` has an `Identifier()` function, but it doesn't return an identifier that's unique to this tailer anymore. That identifier is used for example in the auditor to track offsets. Let's make it clear inline that it's not a unique identifier anymore.
* related to this: now that we can have more than one tailer with the same identifier submitting messages at the same time, we may run into weird race conditions in the auditor that tracks offsets. Maybe we could improve this by stopping the tracking of offsets on the older tailers? (may make the logic much more complex, would require having a second map to track tailers by file path, so not sure it's worth it)
* This is still tracked by `path`, even though tailers are not uniquely identified by `path` anymore: https://github.com/DataDog/datadog-agent/blob/271bea9dfc3b6318900ef227cc4af02edca6954a/pkg/logs/input/file/tailer.go#L126 I think this could lead to race conditions as well, and lead to missing sources in some cases, could you have a look and evaluate the impact?

**Comment by remeh** [comment]
@olivielpeau thanks for your detailed review, that's very helpful.

> * We focused on the impact of this change on tailed containers, but not really on regular tailed log files. With this change, I guess we could now have more than one tailer for a given log file if this log file is configured to be tailed from more than one integration config file, or am I missing something? This sounds like an undesirable change that could double (or more) the amount of logs sent in these cases.

The code executed for the scanning of files in a directory for a integration config has not been changed, except for the key that is used in the map. Because the `getSourceIdentifier()` is empty in this case, the key will be `/path/to/file.log/` (instead of `/path/to/file.log` before) and it won't start two tailers, even if two different configurations mention this file.
I've pushed a change for `buildTailerKey` to return `/path/to/file.log` (without the last `/`) for correctness.

However, we could wonder how it is stored in the registry: if I'm correct, it is using the `Identifier()` method, thus, nothing has been changed on this side.

I've also locally compared an Agent 7.22.0 and an Agent with this patch and I don't observe difference in neither the status, the logs, nor the logs visible in the app.

> * Tailer has an Identifier() function, but it doesn't return an identifier that's unique to this tailer anymore. That identifier is used for example in the auditor to track offsets. Let's make it clear inline that it's not a unique identifier anymore.

Correct, I've improved the `Tailer.Identifier()` comment.

> * related to this: now that we can have more than one tailer with the same identifier submitting messages at the same time, we may run into weird race conditions in the auditor that tracks offsets. Maybe we could improve this by stopping the tracking of offsets on the older tailers? (may make the logic much more complex, would require having a second map to track tailers by file path, so not sure it's worth it)

At this point, I think that would be better handled during a refactor / improvements PR.  WDYT?

> * This is still tracked by path, even though tailers are not uniquely identified by path anymore:

The list of inputs of a Tailer's sources is only used to build the status, I don't think it will be an issue and it's better to display the filepath here:
  - https://github.com/DataDog/datadog-agent/blob/271bea9dfc3b6318900ef227cc4af02edca6954a/pkg/logs/status/builder.go#L129
  - https://github.com/DataDog/datadog-agent/blob/271bea9dfc3b6318900ef227cc4af02edca6954a/pkg/status/templates/logsagent.tmpl#L64

This is not clear and something which is in a list I've started of feasible refactor improvements.

File: pkg/logs/input/file/tailer_key.go
Comment (remeh) [reply pkg/logs/input/file/tailer_key.go:13]: What do you think of:

```
type tailerKey interface {
	// getFilepath returns the filepath of the tailed file. Should never be empty.
	getFilepath() string
	// getContainerID returns the Container ID when the tailer is tailing a log file from a container. Could be empty.
	getContainerID() string
}
```

All these abstractions in the Logs Agent is not always benefiting the dev.

Comment (remeh) [reply pkg/logs/input/file/tailer_key.go:13]: Nevermind this comment. I've refactored File into Tailer and this is no longer needed.

File: pkg/logs/input/file/tailer_key.go
Comment (remeh) [reply pkg/logs/input/file/tailer_key.go:11]: I agree that this would be the way to go (Tailer should contain File, they have a hard dependency in this case and that's way easier to think this way).

I've done it in a branch and overall I feel like the result is fine, slightly more clear but I'm afraid it's too big of a change if we want to ship this tomorrow, see the commit here: https://github.com/DataDog/datadog-agent/commit/59bd1969fd12c98bd13e1c29e2ebc698dd353feb It introduces the concept the the Scanner needs a key, which is way more clear to me than mentioning some source configs.

This is one in many refactor that we should consider within these sources, however being this large, I would vote for pushing this into master later-ish.

**Review by olivielpeau — APPROVED** [review]
LGTM, thanks! This PR is a tricky one. You may (or may not) want to do a small renaming on the functions of the `tailerKey` interface before merging, but in any case LGTM 👍 

Interested in reading the list of follow-ups you'll have for this, when it's ready :)  (from a functional/behavior perspective, the follow-up on the auditor race condition may be the top priority one, since there's clearly a race condition happening there IMO)

> The code executed for the scanning of files in a directory for a integration config has not been changed, except for the key that is used in the map. Because the getSourceIdentifier() is empty in this case, the key will be /path/to/file.log/ (instead of /path/to/file.log before) and it won't start two tailers, even if two different configurations mention this file.
I've pushed a change for buildTailerKey to return /path/to/file.log (without the last /) for correctness.

Ok thanks for checking and explaining. I was also thinking of the case of a logs config of type `file` defined as a pod annotation/docker label on a container (which would both have a container ID defined under `Source.Config.Identifier` and be a more "regular" tailed log file), but [1] I don't even know if it makes sense for a user to define such configs and [2] the new behavior introduced by the PR is probably better anyway. In any case, we're good here.

> At this point, I think that would be better handled during a refactor / improvements PR. WDYT?

Sounds good to me.

> The list of inputs of a Tailer's sources is only used to build the status

Yes, and re-reading the code, I realize the source in `t.source.AddInput(t.path)` would actually be a different source for the 2 containers, so there's no race condition or issue here.

**Pull request merged**
